packages/common/src/search/Catalog.ts

```diff
@@ -0,0 +1,198 @@
+   * '''js
+   * const catalog = await Catalog.create('https://site-org.hub.arcgis.com', context);
+   * '''
+   *
+   * @param identifier
+   * @param context
+   * @returns
+   */
+  public static async init(
+    identifier: string,
+    context: IArcGISContext
+  ): Promise<Catalog> {
+    // fetch the catalog
+    const fetched = await fetchCatalog(identifier, context.hubRequestOptions);
+    // return an instance
+    return new Catalog(fetched, context);
+  }
+
+  /**
+   * Create a Catalog instance from a Catalog Json object
+   * @param json
+   * @param context
+   * @returns
+   */
+  static fromJson(json: any, context: IArcGISContext): Catalog {
+    // ensure it's in the latest structure
+    json = upgradeCatalogSchema(json);
+    return new Catalog(cloneObject(json), context);
+  }
+
+  /**
+   * Return the JSON object backing the instance
+   * @returns
+   */
+  toJson(): IHubCatalog {
+    return cloneObject(this._catalog);
+  }
+
+  /**
+   * Return the schema version
+   */
+  get schemaVersion() {
+    return this._catalog.schemaVersion;
+  }
+
+  /**
+   * Title getter
+   */
+  get title() {
+    return this._catalog.title;
+  }
+
+  /**
+   * Titke setter
+   */
+  set title(v: string) {
+    this._catalog.title = v;
+  }
+
+  /**
+   * Return the existing scopes hash
+   */
+  get scopes(): ICatalogScope {
+    return this._catalog.scopes;
+  }
+
+  /**
+   * Get the scope's query for a particular entity type
+   * @param type
+   * @returns
+   */
+  getScope(type: EntityType): IQuery | undefined {
+    return this._catalog.scopes[type];
+  }
+
+  /**
+   * Set the scope for a specific entity type
+   * @param type
+   * @param query
+   */
+  setScope(type: EntityType, query: IQuery) {
+    this._catalog.scopes[type] = query;
+  }
+
+  /**
+   * Get the collections array. Returns simple objects not Collection instances
+   */
+  get collections() {
+    return this._catalog.collections || [];
+  }
+
+  /**
+   * Get the names of the collections
+   */
+  get collectionNames(): string[] {
+    return mapBy("key", this.collections);
+  }
+
+  /**
+   * Get a Collection instance by name
+   * @param name
+   * @returns
+   */
+  getCollection(name: string): Collection {
+    const json = this._catalog.collections.find((entry) => entry.key === name);
+    if (json) {
+      // clone it then merge in the associated scope filter
+      const clone = cloneObject(json);
+      const catalogScope = this.getScope(clone.scope.targetEntity);
+      if (catalogScope?.filters) {
+        clone.scope.filters = [...clone.scope.filters, ...catalogScope.filters];
+      }
+      return Collection.fromJson(clone, this._context);
+    } else {
+      throw new HubError(
+        "getCollection",
+        `Collection "${name}" is not present in the Catalog`
+      );
+    }
+  }
+
+  /**
+   * Execute a search against the Catalog as a whole
+   * @param query
+   * @param targetEntity
+   * @returns
+   */
+  search(
```

Comment:
I wonder if instead of `i.search(query, { targetEntity: "groups" })` we should just have `i.searchItems(query)` and `i.searchGroups(query)`?

Reply:
The intent here is that we should be able to use this single fn to search for any of those entities. That said, needing to pass in the entityType in the searchOptions seems confusing.

At one point AT had wanted a catalog search to execute a query for each entity that had a scope, so you'd get an array if `IHubSearchResponse<IHubSearchResult>` but in later discussions re: the http api, I understood that it would specify the entity type as a url segment, and thus return a single `IHubSearchResponse<IHubSearchResult>`.

Anyhow - I'm open to changing this, but that's the back story

Reply:
This is a fine implementation now but will definitely get more complex with general searching and "common" filters. For example, if one provides an IQuery with a single `title` filter, and the catalog's scope contains multiple targetEntity types (e.g. `item` and `group`) it's not unreasonable to assume it would be able to search across those multiple scopes simultaneously. The complexity is then determining which filters are "allowed" for each given  type.

Reply:
@tomwayson  & @drspacemanphd - I'm now leaning towards having separate functions, as that simplifies the API - want to search for Groups? use `myCatalog.searchGroups(...)`. 

Another related question is how to handle catalog level queries for entities that don't have a scope defined. i.e. if a catalog only defines `item` scope, and we search for `group` - what should happen? I'm leaning towards an exception.

As for searching across the entity types, I think that leads to even more complexity (how to handle paging, aggregations etc etc) so my inclination is to say that's just not supported. While it would be "neat" to issue a single query and get back users, groups, events and items related to a single search term, I don't think that's a critical feature, and any UI that wants to show that information could manage it via multiple search calls.